text/3678-trait-method-impl-restrictions.md

```diff
@@ -0,0 +1,145 @@
+implementations, which become available on every implementation of the trait.
+However, the implementer of the trait can still provide their own
+implementation of such a method. In some cases, the trait does not want to
+allow implementations to vary, and instead wants to guarantee that all
+implementations of the trait (or all third-party implementations outside the
+module or crate) use an identical method implementation. For instance, this may
+be an assumption required for correctness.
+
+This RFC allows restricting the implementation of trait methods.
+
+Alternatively, a trait may wish to allow implementations within the same crate
+to override a method, but not allow external implementations to override that
+method.
+
+This mechanism also faciliates marker-like traits providing no implementable
+methods, such that implementers only choose whether to provide the trait and
+never how to implement it; the trait then provides all the method
+implementations.
+
+One example of a trait in the standard library benefiting from this:
+`Error::type_id`, which has thus far remained unstable because it's unsafe to
+override. This RFC would allow stabilizing that method so users can call it,
+without permitting reimplementation of it.
+
+# Explanation
+[explanation]: #explanation
+
+When defining a trait, the definition can annotate methods or associated
+functions to restrict whether implementations of the trait can define them. For
+instance:
+
+```rust
+trait MyTrait: Display {
+    impl(crate) fn method(&self) {
+        println!("MyTrait::method: {self}");
+    }
+}
+```
+
+Note that if a method or associated function marked with an `impl` restriction
+does not have a default body, the trait will not be possible to implement
+outside the indicated visibility. For instance, if a trait has a method with
+`impl(crate)` and no default body, the trait will not be possible to implement
+outside the crate.
+
+When implementing a trait, the compiler checks if any implemented methods or
+associated functions have an `impl` restriction. If so, and the restriction
+does not include the current module, the compiler will emit an error on the
+implementation. If the method or associated function has a default
+implementation in the trait, the compiler will suggest removing the
+implementation from the `impl Trait for` block; otherwise, the compiler will
+state that the trait cannot be implemented here because the method or
+associated function cannot be implemented here.
+
+In every other way, an `impl`-restricted method or associated function acts
+identically to any other method or associated function, and can be invoked
+accordingly:
+
+```rust
+fn takes_mytrait(m: &impl MyTrait) {
+    m.method();
+}
+```
+
+Note that in some cases, the compiler may be able to avoid placing an
+`impl`-restricted method in the trait's vtable, if there is only a single
+implementation that can ever be invoked.
```

Comment:
@scottmcm I don't think I understand how that case relates. We can't do this for `RangeBounds` because there are multiple implementations of `RangeBounds`.

If we have a trait where there can only be one implementation of a given method, that effectively lets us do a trivial version of the devirtualization optimization: call the one and only implementation directly, no indirection.

Reply:
@scottmcm I see; thanks for the explanation. Based on that example, I agree that there are cases where the compiler may prefer to monomorphize a method and put it into the vtable.

I think it's still useful for the RFC to observe that the compiler *may* wish to devirtualize, if the compiler decides that the one-and-only implementation *doesn't* benefit from monomorphization. But I'll reword it to clarify and avoid the implication that doing so would always be the right thing if possible.